lsp-mode.el

```diff
@@ -6039,7 +6060,24 @@ When UPDATE? is t force installation even if the server is present."
 (defvar lsp-deps-providers
   (list :npm (list :path #'lsp--npm-dependency-path
                    :install #'lsp--npm-dependency-download)
-        :system (list :path #'executable-find)))
+        :system (list :path #'lsp--system-path)))
+
+(defun lsp--system-path (path)
+  "If PATH is absolute and exists return it as is. Otherwise,
+return the absolute path to the executable defined by PATH or
+nil."
+  ;; For node.js 'sub-packages' PATH may point to a *.js file. Consider the
+  ;; typescript-language-server. When lsp invokes the server, lsp needs to
+  ;; supply the path to the typescript compiler, tsserver.js, as an argument. To
+  ;; make code platform independent, one must pass the absolute path to the
+  ;; tsserver.js file (Windows requires a *.js file - see help on the JavaScript
+  ;; child process spawn command that is invoked by the
+  ;; typescript-language-server). This is why we check for existence and not
+  ;; that the path is executable.
+  (or (and (f-absolute? path)
```

Comment:
nit: this can be rewrite in `if` term, which is easier to understand.

Reply:
I suggested that version and I usually favour the functional version over the imperative one, e. g. 
```
(or (foo) (bar) (foo-bar) )
```
over
```
(if-let (foo (foo))
    foo
  (if-let (bar (bar))
      bar
    (foo-bar)))
```

Reply:
I agree with your case of writing (using `if-let`), the `if-let` look overly verbose, however in this case, the functional version of 
``` elisp
(or (and (f-absolute? path)
         (f-exists? path)
         path)
    (executable-find path))
```
can be reduce to 
``` elisp
(if (and (f-absolute? path)
         (f-exists? path))
    path
  (executable-find path))
```
Which doesn't riddle me of shouldn't the last `path` in `and` clause being check before used in `f-absolute?`

Reply:
I agree for that particular case - `if` statement looks fine too.

> Which doesn't riddle me of shouldn't the last path in and clause being check before used in f-absolute?

Can you elaborate? You mean like that 
```
(and path
         (f-absolute? path)
         (f-exists? path)
         path)
```

Reply:
I means when I look at
``` elisp
(and (f-absolute? path)
         (f-exists? path)
         path)
```
First thought come to me as: "Oh, they're checking for path nullness too". Then "wait why don't path nullness being checked before use?". Then I finally realize that they just want to return `path` here, since if path is `nil`, the another clause of `or` will throw too

Reply:
I updated the code to be 

```emacs-lisp
(if (and (f-absolute? path)
         (f-exists? path))
    path
  (executable-find path))
```
which to me is a little easier to parse because lisp isn't my native coding language, though the functional version is also fairly clean to me.